src/hooks.js

```diff
@@ -0,0 +1,77 @@
+import {callback, defined} from 'chart.js/helpers';
+
+export const elementHooks = ['afterDraw', 'beforeDraw'];
+export const hasLabel = (el) => el && 'drawLabel' in el && el.options && el.options.label;
+
+export function updateHooks(chart, state, options) {
+  const visibleElements = state.visibleElements;
+  state.hooked = false;
+
+  elementHooks.forEach(hook => {
+    if (typeof options[hook] === 'function') {
+      state.hooked = true;
+      state.hooks[hook] = options[hook];
+    } else if (defined(state.hooks[hook])) {
+      delete state.hooks[hook];
+    }
+  });
+
+  if (!state.hooked) {
+    visibleElements.forEach(scope => {
+      if (!state.hooked) {
+        elementHooks.forEach(hook => {
+          if (typeof scope.options[hook] === 'function') {
+            state.hooked = true;
+          }
+        });
+      }
+    });
+  }
+}
+
+export function drawElement(chart, state, element) {
+  draw(chart, state, element, {method: 'draw', status: '_drawElement'});
+}
+
+export function drawLabelElement(chart, state, element) {
+  draw(chart, state, element, {method: 'drawLabel', status: '_drawLabel'});
+}
+
+function draw(chart, state, element, options) {
+  if (beforeDraw(state, element) === false) {
+    return;
+  }
+  const method = element[options.method];
+  method.call(element, chart.ctx, chart.chartArea);
+  element[options.status] = true;
+  afterDraw(state, element);
+}
+
+function beforeDraw(state, element) {
+  if (element._draw === false) {
+    return false;
+  }
```

Comment:
why is this needed?

Reply:
The _draw is needed because element and label are drawn in 2 different time frame. Therefore if the hook returns false (for instance before drawing the element), the label shouldn't be drawn, when invoked.

Reply:
the `_draw` property has been removed